### DISCUSSION TITLE
iOS application packaging failed

### DISCUSSION THREAD

**Discussion by ichark** [airsdk/Adobe-Runtime-Support#1830] [Q&A] [answered]
Hi!  
I have some strange errors when I try to package the iOS production build.
Our previous .p12 certificate has expired. 
After generating a new .p12 certificate I have had the next error. 
I tried to use a different AIR SDK from 33.1.1.554. to  33.1.1.795 but the result same.

```
Failed to package AIR application ios_release.ipa:
Error: Can't replace shorter signature with longer one

```

Could anyone have the same issue or have some advice for this situation.

**Comment by ajwfrost** [comment]
That error is happening if there's a binary file with a code signature section in it already, and if we're then trying to generate the new code signature block, but where there's just not enough room in the original section for the updated signature to fit...

Can I check what operating system/version you're using? It may be that you can get around this on a mac if you set it to use the native codesign process instead (see the AIR SDK "lib" folder, "adt.cfg" configuration file, and uncomment `UseNativeCodesign=true`)

Are you using AIR Native Extensions? as I'm wondering if it's one of these which is already signed where we're then trying to re-sign it. You can also enable logging in that adt.cfg file (`DebugOut=true`) and then check the log which is in the home folder (/Users/username, or c:\users\username on Windows) as it may tell us what's being signed..

thanks

**Comment by ichark** [comment, marked as answer]
Hi @ajwfrost,  
Thanks for your help!  
I'm tried making builds on both platforms win and mac. 
I'm using ANE from Distriqt and seems the problem was in Facebook Frameworks.  
On another project with non-updated Facebook SDK and a new .p12 certificate distribution app packaged successfully.  
Here are 2 logs from adl.log

```
Finalising signature for framework: Frameworks/FBAEMKit.framework/FBAEMKit, file is D:\Work\GIT\EasterSweeper\Frameworks\FBAEMKit.framework\FBAEMKit
Config type ipa-app-store, platform is iOS
SwiftSupport files: copied as-is
Swift libraries in payload: copied as-is
Checking certificate chain for the issuer
Signing certificate subject name = C=BE,O=SmileyGamer BVBA,OU=********,CN=Apple Distribution: SmileyGamer BVBA (********),UID=********
Signing certificate issuer name  = C=US,O=Apple Inc.,OU=G3,CN=Apple Worldwide Developer Relations Certification Authority
Found certificate in chain, subject name = C=BE,O=SmileyGamer BVBA,OU=********,CN=Apple Distribution: SmileyGamer BVBA (********),UID=********
Found certificate in chain, subject name = CN=Apple Worldwide Developer Relations Certification Authority,OU=Apple Worldwide Developer Relations,O=Apple Inc.,C=US
Found certificate in chain, subject name = C=US,O=Apple Inc.,OU=G3,CN=Apple Worldwide Developer Relations Certification Authority
Match -> found issuer
..java.base/java.lang.Thread.getStackTrace(Thread.java:1606)
..com.adobe.air.DebugFile.printStack(DebugFile.java)
..com.adobe.air.ADTException.<init>(ADTException.java)
..com.adobe.air.ipa.MachoSigner.sign(MachoSigner.java)
..com.adobe.air.ipa.MachoSigner.sign(MachoSigner.java)
..com.adobe.air.ipa.IPASigner.signMacho(IPASigner.java)
..com.adobe.air.ipa.IPAOutputStream.finalizeFrameworkSig(IPAOutputStream.java)
..com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
..com.adobe.air.ipa.IPAPackager.createPackage(IPAPackager.java)
..com.adobe.air.ADT.parseArgsAndGo(ADT.java)
..com.adobe.air.ADT.run(ADT.java)
..com.adobe.air.ADT.main(ADT.java)
Stack trace:
ADT exception: Error: Can't replace shorter signature with longer one  
```

and log from successful packaging
```
Finalising signature for framework: Frameworks/FBSDKCoreKit.framework/FBSDKCoreKit, file is D:\Work\GIT\EasterSweeper\Frameworks\FBSDKCoreKit.framework\FBSDKCoreKit
Config type ipa-app-store, platform is iOS
SwiftSupport files: copied as-is
Swift libraries in payload: copied as-is
Checking certificate chain for the issuer
Signing certificate subject name = C=BE,O=SmileyGamer BVBA,OU=********,CN=Apple Distribution: SmileyGamer BVBA (********),UID=********
Signing certificate issuer name  = C=US,O=Apple Inc.,OU=G3,CN=Apple Worldwide Developer Relations Certification Authority
Found certificate in chain, subject name = C=BE,O=SmileyGamer BVBA,OU=********,CN=Apple Distribution: SmileyGamer BVBA (********),UID=********
Found certificate in chain, subject name = CN=Apple Worldwide Developer Relations Certification Authority,OU=Apple Worldwide Developer Relations,O=Apple Inc.,C=US
Found certificate in chain, subject name = C=US,O=Apple Inc.,OU=G3,CN=Apple Worldwide Developer Relations Certification Authority
Match -> found issuer
Finalising signature for framework: Frameworks/FBSDKLoginKit.framework/FBSDKLoginKit, file is D:\Work\GIT\EasterSweeper\Frameworks\FBSDKLoginKit.framework\FBSDKLoginKit
Config type ipa-app-store, platform is iOS
SwiftSupport files: copied as-is
Swift libraries in payload: copied as-is
Finalising signature for framework: Frameworks/FBSDKShareKit.framework/FBSDKShareKit, file is D:\Work\GIT\EasterSweeper\Frameworks\FBSDKShareKit.framework\FBSDKShareKit
Config type ipa-app-store, platform is iOS
SwiftSupport files: copied as-is
Swift libraries in payload: copied as-is
Finalising signature for framework: Frameworks/libswiftCore.dylib, file is D:\Work\GIT\EasterSweeper\Frameworks\libswiftCore.dylib
Config type ipa-app-store, platform is iOS
SwiftSupport files: copied as-is
Swift libraries in payload: copied as-is

Copying the input file (swift library) into the payload as-is
```

**Reply by SeanRuanTEC** [reply]
I've got the same issues
AIR SDK v33.1.1.821
Distriqt's Facebook Ane v10.0.1

**Reply by ChunYuan-Wang** [reply]
Is it solved? I got the same issue.

**Reply by ajwfrost** [reply]
If you're running on macOS, can you try setting up the configuration to use the "native" codesign tool for code signing? i.e. via the AIR SDK folder then "lib/adt.cfg", uncomment `UseNativeCodesign=true`

Otherwise, I don't know that we're able to do anything here, but as a workaround it may be that you can manually code-sign that framework file prior to it being used by ADT? (although if you're on Windows of course, that doesn't help..)

Let me double-check, perhaps we can see what the differences are and whether we can modify things to free up the necessary space for the new code signature format...

thanks

**Reply by Kelsonk** [reply]
Any updates? I'm also getting this issue after not updating my iOS game in awhile and I don't know where to start. Just trying to build on Windows. Neither Air 33 or Air 50 is working.

**Reply by spielepalast** [reply]
any updates?

**Reply by ajwfrost** [reply]
Hi

We'd missed this one, sorry ... but looking at it -> the `FBAEMKit` file doesn't appear to be signed at all, either from Facebook or from Distriqt, which should mean that it's impossible to get the "Can't replace shorter signature with longer one" error....

Is it possible to confirm that this is still the error message, and if so can we please get a copy of the file that's causing the error (i.e. from the previous "Finalising signature for..." line)?

thanks

**Reply by Kelsonk** [reply]
I don't know what that is, but this has gotten really painful now that's it's happening with my debug build. I used to only have to build the distribution build on a Mac and was able to test and build a dev build on my PC, but not anymore

**Reply by ajwfrost** [reply]
@Kelsonk is there any chance you can share the extension/framework that's failing so that we can inspect this and try to reproduce it here? Link for upload: https://transfer.harman.com/requests/iTQZ4NmGVXUy2RUOa3ty2F
thanks

**Reply by Kelsonk** [reply]
I don't know how to determine that

**Reply by ajwfrost** [reply]
@Kelsonk probably to start with, we need to check it's the same issue. So if you can edit (or create) a file in your `c:\users\username\.airsdk` folder called `adt.cfg` and add a line `DebugOut=true`, then try to reproduce the error. Once you've run the build, there should be a log file available: `c:\users\username\adt.log`

The log should have a number of lines like:
```
Finalising signature for framework: .....
```
and then at some point:
```
ADT exception: Error: Can't replace shorter signature with longer one 
```

So then, we're looking for the last "framework" mention before that exception is thrown. Hopefully/probably, this is still `Frameworks/FBAEMKit.framework/FBAEMKit`. And if so, it should say where the actual file is, and that's the file I'd like to get hold off...

thanks

**Comment by Kelsonk** [comment]
Yeah, I'll send you the file.

Stack trace:
ADT exception: Error: Can't replace shorter signature with longer one 
*** ADT Logging at 2023/02/28 11:02:21 ***
Called by: C:\air\lib\adt.jar
ADT 33.1.1.929 called with: -package -target ipa-debug -connect 192.168.5.208 -storetype pkcs12 -keystore cert\8-30-2022\2272023.p12 -storepass xxxxxx -provisioning-profile cert\8-30-2022\2272023dev.mobileprovision dist\ClickerNewIos-debug.ipa applicationapm.xml -C binAssets . -C biniOS . -C apm/assets/ios . -C icons/ios . -extdir apm/ane
*** ADT Logging at 2023/02/28 11:02:22 ***
Called by: C:\air\lib\adt.jar
ADT 33.1.1.929 called with: -package -target ipa-debug -connect 192.168.5.208 -storetype pkcs12 -keystore cert\8-30-2022\2272023.p12 -storepass xxxxxx -provisioning-profile cert\8-30-2022\2272023dev.mobileprovision dist\ClickerNewIos-debug.ipa applicationapm.xml -C binAssets . -C biniOS . -C apm/assets/ios . -C icons/ios . -extdir apm/ane
Warning: Using default implementation for native extension 'com.google.android.material'
Warning: Using default implementation for native extension 'androidx.constraintlayout'
Warning: Using default implementation for native extension 'androidx.cardview'
Warning: Using default implementation for native extension 'com.google.code.gson'
Warning: Using default implementation for native extension 'androidx.appcompat'
Warning: Using default implementation for native extension 'com.android.installreferrer'
Warning: Using default implementation for native extension 'com.distriqt.playservices.AppSet'
Warning: Using default implementation for native extension 'com.jetbrains.kotlin'
Warning: Using default implementation for native extension 'com.distriqt.playservices.Ads'
Warning: Using default implementation for native extension 'androidx.emoji2'
Warning: Using default implementation for native extension 'com.google.android.play'
Warning: Using default implementation for native extension 'com.distriqt.playservices.Analytics'
Warning: Using default implementation for native extension 'androidx.core'
Warning: Using default implementation for native extension 'androidx.browser'
Warning: Using default implementation for native extension 'androidx.work'
Warning: Using default implementation for native extension 'androidx.vectordrawable'
Warning: Using default implementation for native extension 'com.distriqt.playservices.AdsIdentifier'
Warning: Using default implementation for native extension 'com.distriqt.playservices.Base'
Warning: Using default implementation for native extension 'androidx.room'
Finalising signature for framework: Frameworks/FBAEMKit.framework/FBAEMKit, file is C:\Users\kelso\OneDrive\Desktop\castleclickNew\apm\assets\ios\Frameworks\FBAEMKit.framework\FBAEMKit
Config type ipa-debug, platform is iOS
SwiftSupport files: copied as-is
Swift libraries in payload: copied as-is
Checking certificate chain for the issuer
Signing certificate subject name = C=US,O=Kelson Kugler,OU=8GML6J39M4,CN=Apple Development: Kelson Kugler (8578MQ9536),UID=CWWXDSTTNG
Signing certificate issuer name  = C=US,O=Apple Inc.,OU=G3,CN=Apple Worldwide Developer Relations Certification Authority
Found certificate in chain, subject name = C=US,O=Kelson Kugler,OU=8GML6J39M4,CN=Apple Development: Kelson Kugler (8578MQ9536),UID=CWWXDSTTNG
Found certificate in chain, subject name = CN=Apple Worldwide Developer Relations Certification Authority,OU=Apple Worldwide Developer Relations,O=Apple Inc.,C=US
Found certificate in chain, subject name = C=US,O=Apple Inc.,OU=G3,CN=Apple Worldwide Developer Relations Certification Authority
Match -> found issuer
..java.base/java.lang.Thread.getStackTrace(Thread.java:1606)
..com.adobe.air.DebugFile.printStack(DebugFile.java)
..com.adobe.air.ADTException.<init>(ADTException.java)
..com.adobe.air.ipa.MachoSigner.sign(MachoSigner.java)
..com.adobe.air.ipa.MachoSigner.sign(MachoSigner.java)
..com.adobe.air.ipa.IPASigner.signMacho(IPASigner.java)
..com.adobe.air.ipa.IPAOutputStream.finalizeFrameworkSig(IPAOutputStream.java)
..com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
..com.adobe.air.ipa.IPAPackager.createPackage(IPAPackager.java)
..com.adobe.air.ADT.parseArgsAndGo(ADT.java)
..com.adobe.air.ADT.run(ADT.java)
..com.adobe.air.ADT.main(ADT.java)
Stack trace:
ADT exception: Error: Can't replace shorter signature with longer one

**Reply by ajwfrost** [reply]
Perfect, thanks. Are you able to put the adt.jar file from the below zip into your AIR SDK's "lib" folder (please back up the existing one first!) and then try again? If it fails, please let us know what's in this output log, there should be an extra line giving information about the existing signature location and size..

Note that this was built on top of AIR SDK 50.2.1, it should still be compatible with your 33.1.1.929 version, as long as you don't try any Android builds...

[adt.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/10859381/adt.zip)


thanks

**Reply by a9848840** [reply]
Hi , @ajwfrost 

I also have the same problem

After using adt.zip
No error occurred when packaging ipa
But uploading to TestFlight
The following situation occurs

Have an idea?
Air 50.2.1.1


Dear Developer,

We identified one or more issues with a recent delivery for your app, "xxxxxx" x.x.x (x.x.x). Please correct the following issues, then upload again.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/FBAEMKit.framework/FBAEMKit' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/FBSDKCoreKit.framework/FBSDKCoreKit' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/FBSDKCoreKit_Basics.framework/FBSDKCoreKit_Basics' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/FBSDKGamingServicesKit.framework/FBSDKGamingServicesKit' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/FBSDKLoginKit.framework/FBSDKLoginKit' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/FBSDKShareKit.framework/FBSDKShareKit' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/LineSDK.framework/LineSDK' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxxx.app/Frameworks/LineSDKObjC.framework/LineSDKObjC' contains bitcode.

**Reply by ajwfrost** [reply]
Hi @a9848840  - I think this one is actually a completely separate issue...! Are you okay to open this one as a new bug and copy in your TestFlight feedback to it?
thanks

**Reply by a9848840** [reply]
Yes 
After I update FB SDK

Looks like it's related to the Line SDK

sorry 

====================

ITMS-90482: Invalid Executable - The executable 'xxxxx.app/Frameworks/LineSDK.framework/LineSDK' contains bitcode.

ITMS-90482: Invalid Executable - The executable 'xxxxx.app/Frameworks/LineSDKObjC.framework/LineSDKObjC' contains bitcode.

**Reply by marchbold** [reply]
This is an issue with the frameworks packaged in those extensions. Post an issue on the relevant extension GitHub repo and I'll follow up there.

Most have been resolved but the linesdk one isn't part of our normal support list.
